spark/src/main/scala/io/delta/sharing/spark/filters/OpConverter.scala

```diff
@@ -0,0 +1,191 @@
+    if (leafOp.isDefined) {
+      return leafOp.get
+    }
+
+    // Try converting to a non-leaf op.
+    expr match {
+      // Convert boolean logic operators.
+      case SqlAnd(left, right) =>
+        AndOp(Seq(convertOne(left), convertOne(right)))
+      case SqlOr(left, right) =>
+        OrOp(Seq(convertOne(left), convertOne(right)))
+      case SqlNot(child) =>
+        NotOp(Seq(convertOne(child)))
+
+      // Convert comparison operators.
+      case SqlEqualTo(left, right) =>
+        EqualOp(Seq(convertAsLeaf(left), convertAsLeaf(right)))
+      case SqlLessThan(left, right) =>
+        LessThanOp(Seq(convertAsLeaf(left), convertAsLeaf(right)))
+      case SqlLessThanOrEqual(left, right) =>
+        LessThanOrEqualOp(Seq(convertAsLeaf(left), convertAsLeaf(right)))
+      case SqlGreaterThan(left, right) =>
+        GreaterThanOp(Seq(convertAsLeaf(left), convertAsLeaf(right)))
+      case SqlGreaterThanOrEqual(left, right) =>
+        GreaterThanOrEqualOp(Seq(convertAsLeaf(left), convertAsLeaf(right)))
+
+      // Convert null operations.
+      case SqlIsNull(child) =>
+        IsNullOp(Seq(convertAsLeaf(child)))
+      case SqlIsNotNull(child) =>
+        NotOp(Seq(IsNullOp(Seq(convertAsLeaf(child)))))
+
+      // Convert <=> operation.
+      case SqlEqualNullSafe(left, right) =>
+        val leftOp = convertAsLeaf(left)
+        val rightOp = convertAsLeaf(right)
+        val leftIsNullOp = IsNullOp(Seq(leftOp))
+        val rightIsNullOp = IsNullOp(Seq(rightOp))
+        // Either both are null, or none is null and they are equal.
+        OrOp(Seq(
+          AndOp(Seq(leftIsNullOp, rightIsNullOp)),
+          AndOp(Seq(
+            NotOp(Seq(leftIsNullOp)),
+            NotOp(Seq(rightIsNullOp)),
+            EqualOp(Seq(leftOp, rightOp))))
+        ))
+
+      // Unsupported expressions.
+      case _ =>
+        throw new IllegalArgumentException("Unsupported expression during conversion " + expr)
+    }
+  }
+
+  /**
+   * Tries to convert the SQL expression into a json predicate leaf op.
+   * If conversion fails, the function will return None.
+   *
+   * @param sqlTypeOpt if specified overrides the dataType specified in
+   * in the SQL expression. It is used to suppoprt "cast" method, which
+  // forces data type convertion during evaluation.
+   */
+  def maybeConvertAsLeaf(
+      expr: SqlExpression,
+      sqlTypeOpt: Option[SqlDataType] = None): Option[LeafOp] = {
+    val sqlType = sqlTypeOpt.getOrElse(expr.dataType)
+
+    expr match {
+      case attr: SqlAttributeReference =>
+        // This represents a column in partition filtering.
+        Some(ColumnOp(attr.name, convertDataType(sqlType)))
+      case lit: SqlLiteral =>
+        // This represents a constant value to be used in comparisons.
+        Some(LiteralOp(lit.toString, convertDataType(sqlType)))
+      case c: SqlCast =>
+        // A cast operation which triggers type conversion.
+        Some(convertAsLeaf(c.child, Some(c.dataType)))
```

Comment:
Can there be nested casts? Should we return `maybeConvertAsLeaf(c.child, Some(c.dataType))` instead?

Reply:
Yes they can be (see multi level cast test).
We need convertAsLeaf and not the may version, since the child  must convert and return an op (so no option).
This works since cast itself is a leaf.